Automattic/portkey-android
# Publish to s3

This PR adds support for publishing `stories`, `mp4compose` & `photoeditor` artifacts to s3 using our [publish-to-s3 Gradle plugin](https://github.com/Automattic/publish-to-s3-gradle-plugin). It also adds support for plugin DSL which is something we are adding across all our projects.

The PR follows the same approach as the [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2083) that added publish-to-s3 support. If you're unfamiliar with that PR, it's definitely worth having a look over the description to understand some of the decisions.

The publishing tasks are handled in Buildkite and depend on each other. `mp4compose` will be the first one to be published, because it has no dependencies. Then, `photoeditor` will be published which depends on `mp4compose` and then `stories` will be published which depends on both `mp4compose` & `photoeditor`. We use a conditional dependency for these projects so that during publishing we can replace the project dependency block with the binary dependency of the project that was published. This helps verify that the previous artifact was published correctly and the dependency graph has a binary dependency instead of a project block. (which wouldn't work when it's added to a client) More details about this in the [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2083) as previously mentioned.

Plugin DSL changes were straightforward, so there isn't much to explain there.

**To test**
* Smoke test the example app
* Follow the testing instructions in https://github.com/wordpress-mobile/WordPress-Android/pull/15301

Note that I am leaving this PR as a draft because https://github.com/wordpress-mobile/WordPress-Android/pull/15301 needs to be ready to merge before we can merge this in. I'd also like to get this PR reviewed by the library maintainers as well, just in case there was anything that was missed.
